engine/classes/Elgg/Application/ServeFileHandler.php

```diff
@@ -0,0 +1,183 @@
+<?php
+namespace Elgg\Application;
+
+use Elgg\Application;
+use Elgg\Config;
+
+
+/**
+ * File server handler
+ *
+ * @access private
+ *
+ * @package Elgg.Core
+ */
+class ServeFileHandler {
+
+	/** @var Application */
+	private $application;
+
+	/** @var Config */
+	private $config;
+
+	/** @var array */
+	private $server_vars;
+
+	/**
+	 * Constructor
+	 *
+	 * @param Application $app         Elgg Application
+	 * @param Config      $config      Elgg configuration
+	 * @param array       $server_vars Server vars
+	 */
+	public function __construct(Application $app, Config $config, $server_vars) {
+		$this->application = $app;
+		$this->config = $config;
+		$this->server_vars = $server_vars;
+	}
+
+	/**
+	 * Handle a request for a file
+	 *
+	 * @param array $path URL path
+	 * @return void
+	 */
+	public function handleRequest($path) {
+
+		if (!preg_match('~/serve-file/e(\d+)/l(\d+)/d([ia])/c([01])/([a-zA-Z0-9\-_]+)/(.*)$~', $path, $m)) {
+			header("HTTP/1.1 400 Bad Request");
+			exit;
+		}
+
+		list(, $expires, $last_updated, $disposition, $use_cookie, $mac, $path_from_dataroot) = $m;
+
+		if ($expires && $expires < time()) {
+			$this->send403('URL has expired');
+		}
+
+		$etag = '"' . $last_updated . '"';
+		$this->handle304($etag);
+
+		// @todo: change to minimal boot without plugins
+		$this->application->bootCore();
+
+
+		$key = _elgg_services()->siteSecret->get();
+		$hmac_data = array(
+			'expires' => (int) $expires,
+			'last_updated' => (int) $last_updated,
+			'disposition' => $disposition,
+			'path' => $path_from_dataroot,
+			'use_cookie' => (int) $use_cookie,
+		);
+		if ((bool) $use_cookie) {
+			$hmac_data['cookie'] = _elgg_services()->session->getId();
+		}
+		ksort($hmac_data);
+
+		$hmac = elgg_build_hmac($hmac_data);
+		if (!$hmac->matchesToken($mac)) {
+			$this->send403();
+		}
+
+		$dataroot = _elgg_services()->config->get('dataroot');
+		if (empty($dataroot)) {
+			$this->send404();
+		}
+
+		$filenameonfilestore = "{$dataroot}{$path_from_dataroot}";
+
+		if (!is_readable($filenameonfilestore)) {
+			$this->send404();
+		}
+
+		$actual_last_updated = filemtime($filenameonfilestore);
+		if ($actual_last_updated != $last_updated) {
+			$this->send403('URL has expired');
+		}
+
+		$mime = $this->getContentType($path_from_dataroot);
+		header("Content-type: $mime", true);
+
+		$filesize = filesize($filenameonfilestore);
+		header("Content-Length: $filesize", true);
+
+		if ($disposition == 'i') {
+			header("Content-disposition: inline");
+		} else {
+			$basename = basename($filenameonfilestore);
+			header("Content-disposition: attachment; filename='$basename'");
+		}
+
+		if ($expires) {
+			$expires_str = gmdate('D, d M Y H:i:s \G\M\T', $expires);
+		} else {
+			$expires_str = gmdate('D, d M Y H:i:s \G\M\T', strtotime("+3 years"));
+		}
+
+		header('Expires: ' . $expires_str, true);
+
+		$cache_control = $use_cookie ? 'no-cache' : 'public';
```

Comment:
Hm, I don't think I understand cache control enough. I will let it sit for now, and we can add it later.